[Test] Extract most PromQL test code into separate packages

bboreham opened the pull request:
It is undesirable to have the `promql` package contain a lot of test-only code, which depends on `testify` and other utilities.

I created three new packages:
* `promqltest`, the test utilities used by promql tests and by other packages such as `promtool`.
* `promql_test` which is most of the tests for `promql` package itself, in the same directory.
* `almost`, which extracts `almost.Equal` used by test and non-test code.

I left some tests in the `promql` package, because they use a lot of unexported fields.
Possibly these indicate that more new packages should be extracted.

bboreham commented:
I think the main people impacted by this are people using PromQL as a library outside of Prometheus.
And I don't know who they are, broadly.  Maybe @charleskorn, @fpetkovski ?

Since I see I made a typo in one of the git commit headers, I will fix that and force-push.

fpetkovski commented:
We run those tests as a library, but moving them to a separate package sounds like a good idea.

pracucci commented:
> promql_test which is most of the tests for promql package itself, in the same directory.

I don't understand why `promql_test` is needed. Which problem do we solve with it?

bboreham commented:
Without it we have a circular dependency between `promql` and `promql/promqltest`.

pracucci commented:
> Without it we have a circular dependency between promql and promql/promqltest.

I see. I don't know. From a design perspective, I'm not sure this change is a net positive benefit. To make it happen, you had to introduce `promql_test` which looks an anti-pattern to me, and expose some internal PromQL Engine features like the newly introduced `TestXXX()` functions which looks a bad smell.

> I think the main people impacted by this are people using PromQL as a library outside of Prometheus.

To directly answer your question, looking at the changes I don't see any blocker for Mimir.

bboreham commented:
> you had to introduce promql_test which looks an anti-pattern to me

Some people hold that this is actually a best practice.

It is described at https://pkg.go.dev/testing:
> The test file can be in the same package as the one being tested, or in a corresponding package with the suffix "_test".

More discussion:
* https://github.com/golang/go/issues/25223
* https://www.reddit.com/r/golang/comments/twx02x/should_tests_and_mocks_be_in_a_test_package_or/

fpetkovski commented on promql/engine.go:
```diff
@@ -584,6 +586,11 @@ func (ng *Engine) newTestQuery(f func(context.Context) error) Query {
 	return qry
 }
 
+func (ng *Engine) TestEnablePerStepStats()              { ng.enablePerStepStats = true }
```
Is it not possible to set these via engine opts?

bboreham replied:
Yes it is possible, but there wasn't a way to get those options in to `NewTestEngine`.
I've pushed a commit which adds 20 `false` parameters so we can do it that way.
The other way would be for this one test to not call `NewTestEngine`.

bboreham commented:
Also one more export: `KahanSum`, a function only used by a test.  Perhaps should be named `TestKahanSum`.
I think the feature can be tested instead via `sum_over_time`.

charleskorn commented:
> I think the main people impacted by this are people using PromQL as a library outside of Prometheus. And I don't know who they are, broadly. Maybe @charleskorn, @fpetkovski ?

These changes look fine to me, I don't think they'll cause any issues for us.

bboreham commented:
I have now removed all new exports from `promql` in this PR, except `NewTestQuery`.
Also I renamed a new file `engine_internal_test.go` [since I read](https://medium.com/@matryer/5-simple-tips-and-tricks-for-writing-unit-tests-in-golang-619653f90742) that some people use this idiom.

If no further comments I will clean up the commits a little and merge.
